pkg/resource/v1/certconfig/delete.go

```diff
@@ -2,14 +2,43 @@ package certconfig
 
 import (
 	"context"
+	"fmt"
 
+	"github.com/giantswarm/apiextensions/pkg/apis/core/v1alpha1"
+	"github.com/giantswarm/microerror"
 	"github.com/giantswarm/operatorkit/framework"
 )
 
 func (r *Resource) ApplyDeleteChange(ctx context.Context, obj, deleteChange interface{}) error {
 	return nil
 }
 
+// NewDeletePatch is called upon observed custom object deletion. It receives
+// the deleted custom object, the current state as provided by GetCurrentState
+// and the desired state as provided by GetDesiredState. NewDeletePatch
+// analyses the current and desired state and returns the patch to be applied by
+// Create, Delete, and Update functions.
 func (r *Resource) NewDeletePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*framework.Patch, error) {
-	return nil, nil
+	delete, err := r.newDeleteChange(ctx, obj, currentState, desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	patch := framework.NewPatch()
+	patch.SetDeleteChange(delete)
+
+	return patch, nil
+}
+
+func (r *Resource) newDeleteChange(ctx context.Context, obj, currentState, desiredState interface{}) ([]*v1alpha1.CertConfig, error) {
+	currentCertConfigs, err := toCertConfigs(currentState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "debug", "finding out which certconfigs have to be deleted")
+
+	r.logger.LogCtx(ctx, "debug", fmt.Sprintf("found %d certconfigs that have to be deleted", len(currentCertConfigs)))
+
+	return currentCertConfigs, nil
```

Comment:
This does not make sense to me right now. Why are we deleting all `certConfigs` in the current state?
You are creating them by comparing the current and desired state and then adding new ones so the current state matches the desired state.
I would have expected the reverse here: comparing the current to desired state and removing the `certConfigs` ones if they are not in the desired state anymore. Did I totally miss the point here?

Reply:
When `GetDesiredState()` produces always the same desired state (no dynamic behaviour e.g. for scaling) I understand that same applies for deletion. `certconfig` resource's function is to order certificates for guest cluster. When custom object is created or updated, this resource's `GetDesiredState()` creates the same set of `CertConfigs` on every run. `NewUpdatePatch()` then compares current set of certconfigs to desired set and returns create change accordingly.

When `certconfig` gets notified for custom object deletion, my logic tells me to delete all existing `CertConfigs` that match filtering for given custom object which the `currentState` here represents.

Am I on a wrong track here? In situation where `GetDesiredState()` behaves dynamically based on changes in custom object, this would be different. This can also differ if changes were not to be applied against k8s API but e.g. some cloud provider instead. WDYT?

Reply:
Okay, I was not aware that the desired state would be static. So while this resource is not yet deleted, I would assume that it get always reconciled so that the current state matches the static desired state. 
Then what you are proposing makes sense as we should hopefully then on deletion always remove the entire current state which matches the desired state. 

The only issue I could see is, if we update the operator and remove one item from the desired state, then it won't reconcile the current state to reflect this, right?

Reply:
Oh, that's a good catch. That implementation belongs to different function though. As I see this, I need to create `newDeleteChangeForUpdatePatch()` and rename this to `newDeleteChangeForDeletePatch()`. Then in `NewUpdatePatch()` call for `newDeleteChangeForUpdatePatch()` to remove those orphaned resources that we discussed in `NewUpdatePatch()` PR. Is this correct?

Reply:
I would assume so, then we also deal with these orphans directly and don't need to act on them in another way.

I am not sure if that is the proper structure and naming in  operatorkit though. I would guess so, though.

Reply:
That's the naming in `kvm-operator` at least: https://github.com/giantswarm/kvm-operator/blob/master/service/kvmconfig/v6/resource/deployment/delete.go#L88

Cool, I'll submit a separate PR for that refactoring once this gets approval so it's clearer how the update functionality changes.